### PR TITLE
switch all .S files to unified asm, and use a macro to setup compiler and some other misc changes

### PR DESCRIPTION
**Pull request by kilograham**
Misc changes:

* switch some code to use more recent helper methods (e.g. busy_wait_at_least_n_cycles)
* add documentation to host divider header (I had this ages ago and never promoted)
* Some compiler warnings

### CODE REVIEW

File: src/rp2_common/pico_platform/include/pico/asm_helper.S [hunk 1]
```diff
@@ -6,6 +6,14 @@
 
 #include "pico.h"
 
+# note we don'tdo this by default in this file for backwards comaptibility with user code
```
Comment (lurch) [inline src/rp2_common/pico_platform/include/pico/asm_helper.S:9]: missing space between "don't" and "do"

File: src/host/hardware_divider/include/hardware/divider.h [hunk 2]
```diff
@@ -7,116 +7,325 @@
...
-// divides unsigned values a by b... (a/b) returned in low 32 bits, (a%b) in high 32 bits... results undefined for b==0
-static inline uint64_t hw_divider_divmod_u32(uint32_t a, uint32_t b) {
+/*! \brief Do an unsigned HW divide and wait for result
+ *  \ingroup hardware_divider
+ *
+ * Divide \p a by \p b, wait for calculation to complete, return result as a fixed point 32p32 value.
+ *
+ * \param a The dividend
+ * \param b The divisor
+ * \return Results of divide as a 32p32 fixed point value.
```
Comment (Wren6991) [inline src/host/hardware_divider/include/hardware/divider.h:29]: Doesn't seem to be a 32p32 fixed point value, just two different 32-bit values

**Review by peterharperuk — DISMISSED** [review]

**Review by Wren6991 — APPROVED** [review]

**Review by Wren6991 — APPROVED** [review]

**Review by Wren6991 — APPROVED** [review]
Thanks for fixing up the comments

File: src/host/hardware_divider/include/hardware/divider.h [hunk 3]
```diff
@@ -7,116 +7,325 @@
...
+ */
+static inline divmod_result_t hw_divider_divmod_u32(uint32_t a, uint32_t b)  {
     if (!b) return (((uint64_t)a)<<32u) | (uint32_t)(-1); // todo check this
```
[the remark this answers is not in the document]
Comment (kilograham) [reply src/host/hardware_divider/include/hardware/divider.h:32]: not too worried about it in the host code

**Pull request merged**